btree/btree.c

```diff
@@ -3529,164 +3565,561 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
 	}
 }
 
-int64_t btree_del_tick(struct m0_sm_op *smop)
-{
-	struct m0_btree_op 	*bop = M0_AMB(bop, smop, bo_op);
-	//ToDo: Implement complete destroy tick function.
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			return P_ACT;
+/* Delete Operation */
 
-		case P_ACT:
-			return P_DONE;
+/**
+ * This function will get called if there is an underflow at current node after
+ * deletion of the record. Currently, underflow condition is defined based on
+ * record count. If record count is 0, there will be underflow. To resolve
+ * underflow,
+ * 1) delete the node from parent.
+ * 2) check if there is an underflow at parent due to deletion of it's child.
+ * 3) if there is an underflow,
+ *        if, we have reached root, handle underflow at root.
+ *        else, repeat steps from step 1.
+ *    else, return next phase which needs to be executed.
+ *
+ * @param bop will provide all required information about btree operation.
+ * @return int64_t return state which needs to get executed next.
+ */
+static int64_t btree_del_resolve_underflow(struct m0_btree_op *bop)
+{
+	struct td              *tree        = bop->bo_arbor->t_desc;
+	struct m0_btree_oimpl  *oi          = bop->bo_i;
+	int                     used_count  = oi->i_used;
+	struct level           *lev         = &oi->i_level[used_count];
+	bool                    flag        = false;
+	struct slot             node_slot;
+	int                     curr_root_level;
+	struct slot             root_slot;
+	struct nd              *root_child;
+
+	do {
+		lev->l_freenode = true;
+		used_count--;
+		lev = &oi->i_level[used_count];
+		node_del(lev->l_node, lev->l_idx, bop->bo_tx);
+		lev->l_node->n_skip_rec_count_check = true;
+		node_slot.s_node = lev->l_node;
+		node_slot.s_idx  = lev->l_idx;
+		node_done(&node_slot, bop->bo_tx, true);
 
-		default:
-			return 0;
-	}
-}
+		/**
+		 * once underflow is resolved at child by deleteing child node
+		 * from parent, determine next step:
+		 * If we reach the root node,
+		 *      if record count > 1, go to P_FREENODE.
+		 *      if record count = 0, set level = 0, height=1, go to
+		 *         P_FREENODE.
+		 *       else record count == 1, break the loop handle root case
+		 *           condition.
+		 * else if record count at parent is greater than 0, go to
+		 *         P_FREENODE.
+		 *      else, resolve the underflow at parent reapeat the steps
+		 *            in loop.
+		 */
+		if (used_count == 0) {
+			if (node_count_rec(lev->l_node) > 1)
+				flag = true;
+			else if (node_count_rec(lev->l_node) == 0) {
+				node_set_level(lev->l_node, 0);
+				tree->t_height = 1;
+				flag = true;
+			} else {
+				lev->l_node->n_skip_rec_count_check = false;
```

Comment:
Not needed as you are again setting it to true at line 3658.

Reply:
done